service/history/shard/context_impl.go

```diff
@@ -1099,24 +1328,131 @@ func acquireShard(
 		timerMaxReadLevelMap[clusterName] = timerMaxReadLevelMap[clusterName].Truncate(time.Millisecond)
 	}
 
-	executionMgr := shardItem.GetExecutionManager()
+	s.wLock()
+	defer s.wUnlock()
+
+	if generation != s.stateGeneration {
+		return false, errOldGeneration
+	}
+
+	s.shardInfo = updatedShardInfo
+	s.remoteClusterCurrentTime = remoteClusterCurrentTime
+	s.timerMaxReadLevelMap = timerMaxReadLevelMap
+
+	return ownershipChanged, nil
+}
+
+func (s *ContextImpl) acquireShard(generation int) {
+	// Retry for 5m, with interval up to 10s (default)
+	// QUESTION: should the 5m limit go in dynamic config?
+	policy := backoff.NewExponentialRetryPolicy(50 * time.Millisecond)
+	policy.SetExpirationInterval(5 * time.Minute)
+
+	isRetryable := func(err error) bool {
+		if common.IsPersistenceTransientError(err) {
+			return true
+		}
+		// Retry this in case we need to create the shard and race with someone else doing it.
+		if _, ok := err.(*persistence.ShardAlreadyExistError); ok {
+			return true
+		}
+		return false
+	}
+
+	// Remember this value across attempts
+	ownershipChanged := false
+
+	try := func() error {
+		// Initial load of shard metadata
+		var err error
+		ownershipChanged, err = s.loadShardMetadata(ownershipChanged, generation)
+		if err != nil {
+			return err
+		}
+
+		s.wLock()
+		defer s.wUnlock()
+
+		// Check that generation still matches
+		if generation != s.stateGeneration {
+			return errOldGeneration
+		}
+
+		// Try to acquire RangeID lock. If this gets a persistence error, it may call:
+		// transitionLocked(contextRequestStop) for ShardOwnershipLostError:
+		//   This will transition to Stopping right here, and the transitionLocked call at the end of the
+		//   outer function will do nothing, since the state was already changed.
+		// transitionLocked(contextRequestLost) for other transient errors:
+		//   This will do nothing, since state is already Acquiring.
+		err = s.renewRangeLocked(true)
+		if err != nil {
+			return err
+		}
+
+		s.logger.Info("Acquired shard")
+
+		// The first time we get the shard, we have to create the engine. We have to release the lock to
+		// create the engine, and then reacquire it. This is safe because:
+		// 1. We know we're currently in the Acquiring state. The only thing we can transition to (without
+		//    doing it ourselves) is Stopped. In that case, we'll have to stop the engine that we just
+		//    created, since the stop transition didn't do it.
+		// 2. We don't have an engine yet, so no one should be calling any of our methods that mutate things.
+		if s.engine == nil {
+			s.wUnlock()
+			s.maybeRecordShardAcquisitionLatency(ownershipChanged)
+			engine := s.createEngine()
+			s.wLock()
+			if generation != s.stateGeneration {
+				engine.Stop()
+				return errOldGeneration
+			}
+			s.engine = engine
```

Comment:
should just create the engine when creating the shardContext, no need to delay it trying to optimize the cost, the benefit of simplify the code is more important.